### DISCUSSION TITLE
Jint 3 memory inefficiency?

### DISCUSSION THREAD

**Discussion by Ghostbird** [sebastienros/jint#1756] [Q&A] [answered]
We moved from a very recent Jint 3 beta to the released Jint 3 and one of our unit tests failed.

Our `Engine` is limited to a maximum of ~18kiB~ 256kiB of memory and a maximum of 10k statements in one execution. We had a test that verified that an infinite loop would throw a `StatementsCountOverflowException`.

The loop is:
```javascript
while (true)
{
  var x = 1;
}
```

With the Jint 3 release, this test fails because the engine exceeds ~18kiB~ 256kiB of memory usage before it exceeds 10k iterations of the loop.

We've managed to fix our test by removing the variable assignment from the loop.
What could have cause the sudden change in memory allocation behaviour?

**Comment by lahma** [comment]
If you can provide a full working test case reproducing the issue and the version it passed with, the issue might be easier to diagnose.

**Reply by lahma** [reply]
Also was it now using 18,1Kb or like 180KB (is the change actually significant)?

**Reply by Ghostbird** [reply]
Thanks for your prompt reaction. I've made a fix to the original message, the memory limit was 256kiB, not 18kiB (2¹⁸, not 18×2¹⁰). It seems that we accidentally chose the values so that it became an edge case. In Jint _3.0.0-beta-2055_ this test passed, but in Jint 3 it often fails. The original one was a poorly constructed legacy test that probably pushed the memory usage slightly higher, but this is a minimal one that has the same limitations:

```csharp
  [Fact(DisplayName = "StatementLimit")]
  public void StatementLimit()
  => new Engine(options =>
  {
    options.MaxStatements(10000);
    options.LimitMemory(2 << 18);
  }).Execute("while(true) { var x = 1; }");
```

This one mostly succeeds in Jint 3, because it doesn't have the overhead introduced by our old test. If you reduce the memory limit to 2¹⁷, it will almost always fail.

Apparently our test was purely accidentally constructed in such a way that it was really close to hitting the memory limit before the statements limit.

Our test isn't very useful at all, what we actually should have tested was whether the limits from our configuration were properly loaded into the Jint `Engine` by our business logic.

**Reply by Ghostbird** [reply, marked as answer]
Conclusion: Apparently our test was purely accidentally constructed in such a way that it was really close to hitting the memory limit before the statements limit, the move to Jint 3 somehow just pushed it over the limit for most runs.
